com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/handler/RefreshFramesHandler.java

```diff
@@ -0,0 +1,129 @@
+* http://www.eclipse.org/legal/epl-v10.html
+*
+* Contributors:
+*     Microsoft Corporation - initial API and implementation
+*******************************************************************************/
+
+package com.microsoft.java.debug.core.adapter.handler;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+
+import com.microsoft.java.debug.core.AsyncJdwpUtils;
+import com.microsoft.java.debug.core.adapter.IDebugAdapterContext;
+import com.microsoft.java.debug.core.adapter.IDebugRequestHandler;
+import com.microsoft.java.debug.core.protocol.Events.StoppedEvent;
+import com.microsoft.java.debug.core.protocol.Messages.Response;
+import com.microsoft.java.debug.core.protocol.Requests.Arguments;
+import com.microsoft.java.debug.core.protocol.Requests.Command;
+import com.microsoft.java.debug.core.protocol.Requests.RefreshFramesArguments;
+import com.sun.jdi.ObjectCollectedException;
+import com.sun.jdi.ThreadReference;
+
+public class RefreshFramesHandler implements IDebugRequestHandler {
+
+    @Override
+    public List<Command> getTargetCommands() {
+        return Arrays.asList(Command.REFRESHFRAMES);
+    }
+
+    @Override
+    public CompletableFuture<Response> handle(Command command, Arguments arguments, Response response,
+            IDebugAdapterContext context) {
+        RefreshFramesArguments refreshArgs = (RefreshFramesArguments) arguments;
+        String[] affectedRootPaths = refreshArgs == null ? null : refreshArgs.affectedRootPaths;
+        List<Long> pausedThreads = getPausedThreads(context);
+        for (long threadId : pausedThreads) {
+            if (affectedRootPaths == null || affectedRootPaths.length == 0) {
+                refreshFrames(threadId, context);
+                continue;
+            }
+
+            Set<String> decompiledClasses = context.getThreadCache().getDecompiledClassesByThread(threadId);
+            if (decompiledClasses == null || decompiledClasses.isEmpty()) {
+                continue;
+            }
+
+            boolean shouldRefresh = false;
+            for (String classUri : decompiledClasses) {
+                // decompiled class uri is like 'jdt://contents/rt.jar/java.io/PrintStream.class?=1.helloworld/%5C/usr%5C/lib%5C/jvm%5C/
+                // java-8-oracle%5C/jre%5C/lib%5C/rt.jar%3Cjava.io(PrintStream.class'.
+                if (classUri.startsWith("jdt://contents/")) {
+                    String jarName = classUri.substring("jdt://contents/".length());
+                    int sep = jarName.indexOf("/");
+                    jarName = sep >= 0 ? jarName.substring(0, sep) : jarName;
+                    for (String affectedRootPath : affectedRootPaths) {
+                        if (affectedRootPath.endsWith("/" + jarName) || affectedRootPath.endsWith("\\" + jarName)) {
+                            shouldRefresh = true;
```

Comment:
Can we directly call `refreshFrames()` and return here?

Or considering extracting the entire for-loop (start from line 56) to another method and early return when affected path is hit.